glslc/src/dependency_info.cc

```diff
@@ -0,0 +1,129 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dependency_info.h"
+
+#include <fstream>
+#include <iostream>
+#include <set>
+#include <sstream>
+
+#include "file.h"
+#include "libshaderc_util/io.h"
+#include "libshaderc_util/string_piece.h"
+
+namespace glslc {
+
+bool DependencyInfoDumpingHandler::DumpDependencyInfo(
+    std::string compilation_output_file_name, std::string source_file_name,
+    std::string* compilation_output_ptr,
+    const std::set<std::string>& dependent_files) {
+  std::string dep_target_label =
+      GetFinalTargetLabel(compilation_output_file_name, source_file_name);
+  std::string dep_file_name =
+      GetFinalDepdencyFileName(compilation_output_file_name, source_file_name);
+
+  std::stringstream dep_string_stream;
+  // dump label and the source_file_name.
+  dep_string_stream << dep_target_label << ": " << source_file_name;
+  // dump the dependent file names.
+  for (auto& dependent_file_name : dependent_files) {
+    dep_string_stream << " " << dependent_file_name;
+  }
+  dep_string_stream << std::endl;
+
+  if (mode_ == dump_as_compilation_output) {
+    compilation_output_ptr->assign(dep_string_stream.str());
+  } else if (mode_ == dump_as_extra_file) {
+    std::ofstream potential_file_stream_for_dep_info_dump;
+    std::ostream* dep_file_stream = shaderc_util::GetOutputStream(
+        dep_file_name, &potential_file_stream_for_dep_info_dump);
+    *dep_file_stream << dep_string_stream.str();
+    if (dep_file_stream->fail()) {
+      std::cerr << "glslc: error: error writing dependent_files info to output "
+                   "file: '"
+                << dep_file_name << "'" << std::endl;
+      return false;
+    }
+  } else {
+    // mode_ should not be 'not_set', we should never be here.
+    return false;
+  }
+  return true;
+}
+
+std::string DependencyInfoDumpingHandler::GetFinalTargetLabel(
+    const std::string& compilation_output_file_name,
+    const std::string& source_file_name) {
+  if (!user_specified_dep_target_label_.empty()) {
+    return user_specified_dep_target_label_;
+  }
+
+  if (mode_ == dump_as_extra_file && compilation_output_file_name != "-" &&
+      compilation_output_file_name != "a.spv") {
+    // In dump_as_extra_file mode, the target follows the
+    // compilation_output_file_name if it is not '-' (means stdout) or 'a.spv'.
+    return compilation_output_file_name;
+  }
+
+  // In dump_as_compilation_output mode, the target name is always induced from
+  // the source file name, unless there is an user specified one.
+  // In dump_as_extra_file mode, only when a compilation_output_file_name is set
+  // to '-' or 'a.spv', we need to deduce the target name from source file name.
+  std::string source_file_base_name =
```

Comment:
If we needed this, then it should have been passed in as an extra argument from the outside.  We should do this complex computation only once in glslc.